phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UncoveredGlobalIndexRegionScanner.java

```diff
@@ -0,0 +1,396 @@
+        int taskCount = setList.size();
+        TaskBatch<Boolean> tasks = new TaskBatch<>(taskCount);
+        for (int i = 0; i < taskCount; i++) {
+            addTasksForScanningDataTableRowsInParallel(tasks, setList.get(i), startTime);
+        }
+        submitTasks(tasks);
+        if (state == State.SCANNING_DATA_INTERRUPTED) {
+            state = State.SCANNING_DATA;
+        } else {
+            state = State.READY;
+        }
+    }
+
+    /**
+     * A page of index rows are scanned and then their corresponding data table rows are retrieved
+     * from the data table regions in parallel. These data rows are then joined with index rows.
+     * The join is for adding uncovered columns to index rows.
+     *
+     * This implementation conforms to server paging such that if the server side operation takes
+     * more than pageSizeInMs, a dummy result is returned to signal the client that more work
+     * to do on the server side. This is done to prevent RPC timeouts.
+     *
+     * @param result
+     * @return boolean to indicate if there are more rows to scan
+     * @throws IOException
+     */
+    @Override
+    public boolean next(List<Cell> result) throws IOException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
+        boolean hasMore;
+        region.startRegionOperation();
+        try {
+            synchronized (innerScanner) {
+                if (state == State.READY && !indexRowIterator.hasNext()) {
+                    state = State.INITIAL;
+                }
+                if (state == State.INITIAL) {
+                    indexRowCount = 0;
+                    indexRows = new ArrayList();
+                    dataRows = Maps.newConcurrentMap();
+                    indexToDataRowKeyMap = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
+                    state = State.SCANNING_INDEX;
+                }
+                if (state == State.SCANNING_INDEX) {
+                    do {
+                        List<Cell> row = new ArrayList<Cell>();
+                        hasMore = innerScanner.nextRaw(row);
+                        if (!row.isEmpty()) {
+                            if (isDummy(row)) {
+                                result.addAll(row);
+                                return true;
+                            }
+                            lastIndexRowKey = CellUtil.cloneRow(row.get(0));
+                            indexToDataRowKeyMap.put(lastIndexRowKey, indexMaintainer.buildDataRowKey(
+                                    new ImmutableBytesWritable(lastIndexRowKey), viewConstants));
+                            indexRows.add(row);
+                            indexRowCount++;
+                            if (hasMore && (EnvironmentEdgeManager.currentTimeMillis() - startTime)
+                                    >= pageSizeMs) {
+                                getDummyResult(lastIndexRowKey, result);
+                                return true;
```

Comment:
Is it safe to return without changing the state? Someone has to update the state. What if all tasks return?

Reply:
I will add a comment for this as follows : We do not need to change the state, State.SCANNING_INDEX since we will continue scanning the index table after the client drops the dummy requests and then calls the next method on its ResultScanner within ScanningResultIterator